test/model/message_test.dart

```diff
@@ -76,6 +94,341 @@ void main() {
     checkNotified(count: messageList.fetched ? messages.length : 0);
   }
 
+  test('dispose cancels pending timers', () => awaitFakeAsync((async) async {
+    final stream = eg.stream();
+    final store = eg.store();
+    await store.addStream(stream);
+    await store.addSubscription(eg.subscription(stream));
+
+    (store.connection as FakeApiConnection).prepare(
+      json: SendMessageResult(id: 1).toJson(),
+      delay: const Duration(seconds: 1));
+    unawaited(store.sendMessage(
+      destination: StreamDestination(stream.streamId, eg.t('topic')),
+      content: 'content'));
+    check(async.pendingTimers).deepEquals(<Condition<Object?>>[
+      (it) => it.isA<FakeTimer>().duration.equals(kLocalEchoDebounceDuration),
+      (it) => it.isA<FakeTimer>().duration.equals(kSendMessageOfferRestoreWaitPeriod),
+      (it) => it.isA<FakeTimer>().duration.equals(const Duration(seconds: 1)),
+    ]);
+
+    store.dispose();
+    check(async.pendingTimers).single.duration.equals(const Duration(seconds: 1));
+  }));
+
+  group('sendMessage', () {
+    final stream = eg.stream();
+    final streamDestination = StreamDestination(stream.streamId, eg.t('some topic'));
+    late StreamMessage message;
+
+    test('outbox messages get unique localMessageId', () async {
+      await prepare(stream: stream);
+      await prepareMessages([]);
+
+      for (int i = 0; i < 10; i++) {
+        connection.prepare(json: SendMessageResult(id: 1).toJson());
+        await store.sendMessage(destination: streamDestination, content: 'content');
+      }
+      // [store.outboxMessages] has the same number of keys (localMessageId)
+      // as the number of sent messages, which are guaranteed to be distinct.
+      check(store.outboxMessages).keys.length.equals(10);
+    });
+
+    Subject<OutboxMessageState> checkState() =>
+      check(store.outboxMessages).values.single.state;
+
+    Future<void> prepareOutboxMessage({
+      MessageDestination? destination,
+      int? zulipFeatureLevel,
+    }) async {
+      message = eg.streamMessage(stream: stream);
+      await prepare(stream: stream, zulipFeatureLevel: zulipFeatureLevel);
+      await prepareMessages([eg.streamMessage(stream: stream)]);
+      connection.prepare(json: SendMessageResult(id: 1).toJson());
+      await store.sendMessage(
+        destination: destination ?? streamDestination, content: 'content');
+    }
+
+    late Future<void> outboxMessageFailFuture;
+    Future<void> prepareOutboxMessageToFailAfterDelay(Duration delay) async {
+      message = eg.streamMessage(stream: stream);
+      await prepare(stream: stream);
+      await prepareMessages([eg.streamMessage(stream: stream)]);
+      connection.prepare(apiException: eg.apiBadRequest(), delay: delay);
```

Comment:
nit: for better realism, make it an error that's consistent with the request having actually succeeded on the server side:
```suggestion
      connection.prepare(httpException: SocketException('failed'), delay: delay);
```

That way it's consistent with the test cases where a corresponding MessageEvent arrives even though the request failed. That also may help a future reader in understanding why that scenario can even happen (which is important to understand since we have some logic carefully written to specifically handle it).